presto-main/src/main/java/com/facebook/presto/type/TimestampOperators.java

```diff
@@ -134,7 +137,7 @@ public static long castToDate(SqlFunctionProperties properties, @SqlType(Standar
             long millis = date + chronology.getZone().getOffset(date);
             return TimeUnit.MILLISECONDS.toDays(millis);
         }
-        return TimeUnit.MILLISECONDS.toDays(value);
+        return floorDiv(value, MILLISECONDS_PER_DAY);
```

Comment:
If the milliseconds value is negative(prior than 1970-01-01), and do not exactly at midnight, `TimeUnit.MILLISECONDS.toDays(value)` would round forwards and `floorDiv(value, MILLISECONDS_PER_DAY)` would round backwards.

Take test case in `TestTimestamp.testCastFromTimestampToDate()` for example, `cast(timestamp '1969-12-31 05:06:07.234' as date)` would return `date '1970-01-01'` in case we use `TimeUnit.MILLISECONDS.toDays(value)`, and would return `date '1969-12-31'` in case we use `floorDiv(value, MILLISECONDS_PER_DAY)`.

Reply:
Ah, I see. Because the due to the division in java, the standard truncating divide rounds forward when numerator is negative. floorDiv makes sure that we round down. Makes sense now. Thanks!